openerp/osv/expression.py

```diff
@@ -732,6 +732,36 @@ def recursive_children(ids, model, parent_field):
                     return ids + recursive_children(ids2, model, parent_field)
                 return [(left, 'in', recursive_children(ids, left_model, parent or left_model._parent_name))]
 
+        def parent_of_domain(left, ids, left_model, parent=None, prefix='', context=None):
+            """ Return a domain implementing the parent_of operator for [(left,parent_of,ids)],
+                either as a range using the parent_left/right tree lookup fields
+                (when available), or as an expanded [(left,in,parent_ids)] """
+            if left_model._parent_store and (not left_model.pool._init):
+                doms = []
+                for o in left_model.browse(cr, uid, ids, context=context):
+                    if doms:
+                        doms.insert(0, OR_OPERATOR)
+                    doms += [AND_OPERATOR, ('parent_right', '>', o.parent_left), ('parent_left', '<=', o.parent_left)]
+                if prefix:
+                    return [(left, 'in', left_model.search(cr, uid, doms, context=context))]
+                return doms
+            else:
+                def recursive_parent(obj, parent_field):
```

Comment:
I understand you tried to use the same design as for the `child_of` code, but I think the `recursive_parent` function could be replaced by a slightly simpler iterative one, such as (untested code):
```python
    def recursive_parent(record, parent_field):
        ids = [record.id]
        while record[parent_field]:
            record = record[parent_field]
            ids.append(record.id)
        return ids
```
You can assume that `parent_field` exists on the model, otherwise using child_of/parent_of must crash.